<!DOCTYPE HTML>
<html>

<script src="js/scripts.js"></script>
<script src="js/gcse.js"></script>
<script src="js/lightbox-plus-jquery.js"></script>

<head>
  <title>Newt Editor</title>
  <meta name="description" content="Newt Pathway Viewer & Editor" />
  <meta name="keywords" content="Pathway, Biological Networks, Visualization, Editor, Curator, SBGN, Systems Biology Graphical Notation, SIF, Simple Interaction Format" />
  <meta http-equiv="content-type" content="text/html; charset=windows-1252" />
  <link rel="icon" href="style/newt-icon-blue.png">
  <link href="style/lightbox.css" rel="stylesheet">
  <link rel="stylesheet" type="text/css" href="style/style.css" title="style" />
</head>

<body>
  <div id="main">
    <div id="header">
      <script> writeLogo(); </script>
      <script> writeMenubar(4); </script>
    </div>
    <div id="site_content">
      <div class="sidebar">
      <script> writeNews(); </script>
      </div>
      <div id="content">
        <h1 id="top-of-tutorials">Tutorials</h1>

        <ul>
        <li> <a href="#introduction">Newt 4</a>
          <span style="float: right;">Updates video: <a href="tutorials-newt4.html"><img src="images/play-button.svg" width=16px/></a></span>
        <li> <a href="#introduction">Introduction</a> 
             <span style="float: right;">Newt basics video: <a href="tutorials-basics.html"><img src="images/play-button.svg" width=16px/></a></span>
        <li> <a href="#menubar-and-toolbar">Menubar and toolbar</a>
        <li> <a href="#right-panel">Right panel</a>
            <span style="float: right;">Newt advanced video: <a href="tutorials-advanced.html"><img src="images/play-button.svg" width=16px/></a></span>
          <ul style="padding-top: 10px;">
          <li style="line-height: 1em;"><a href="#inspector"><i>Object properties inspector and editor</i></a><br>
          <li style="line-height: 1em;"><a href="#map-properties"><i>Map properties</i></a><br>
          <li style="line-height: 1em;"><a href="#experiment-data"><i>Experiment data</i></a><br>
          <li style="line-height: 1em;"><a href="#color-scheme"><i>Color scheme</i></a><br>
          <li style="line-height: 1em;"><a href="#palette"><i>Palette</i></a>
          </ul>
        <li> <a href="#file-operations">File operations</a>
          <ul style="padding-top: 10px;">
          <li style="line-height: 1em;"><a href="#import-export"><i>Import-export</i></a><br>
          <li style="line-height: 1em;"><a href="#static-images"><i>Static images</i></a><br>
          <li style="line-height: 1em;"><a href="#persisting-preferences"><i>Persisting user preferences</i></a><br>
          <li style="line-height: 1em;"><a href="#remote-launch"><i>Launch with a remote SBGN map</i></a><br>
          </ul>
        <li> <a href="#convenient-editing">Convenient editing via SBGN bricks</a>
        <li> <a href="#alignment-grid-and-guidelines">Alignment, grid, and guidelines</a>
        <li> <a href="#semantic-validation">Semantic validation</a>
        <li> <a href="#complexity-management">Complexity management</a>
             <span style="float: right;">Complexity management video: <a href="tutorials-compl.html"><img src="images/play-button.svg" width=16px/></a></span>
          <ul style="padding-top: 10px;">
          <li style="line-height: 1em;"><a href="#hide-show"><i>Hide-show</i></a><br>
          <li style="line-height: 1em;"><a href="#expand-collapse"><i>Expand-collapse</i></a><br>
          </ul>
        <li> <a href="#search-and-highlight">Search and highlight</a>
        <li> <a href="#automatic-layout">Automatic layout</a>
             <span style="float: right;">Automatic layout video: <a href="tutorials-layout.html"><img src="images/play-button.svg" width=16px/></a></span>
        <li> <a href="#querying-pathway-databases">Querying Pathway Databases</a>
        <li> <a href="#references">References</a>
        </ul>

        <h2 id="introduction">Introduction
          <a href="#top-of-tutorials"><img src="images/back-to-top.svg" height=16px/></a></h2>

        <p> Newt editor is composed of a menu bar and a toolbar laid out horizontally on top of the drawing canvas with pan-zoom controls in its upper left, and a right panel for object inspection, map configuration and the palette. </p>

        <span class="center"><a href="images/newt-ss.png" data-lightbox="image-intro" data-title="Overall look and components of the Newt editor"><img src="images/newt-ss.png" style="width: 100%;" /></a></span>

        <p> Newt is a viewer and editor of maps drawn in Systems Biology Graphical Notation (SBGN) Process Description (PD) and 
          Activity Flow (AF) languages, <a target="_blank" href="https://sbml.org/">Systems Biology Markup Language (SBML)</a> or 
          <a target="_blank" href="https://www.pathwaycommons.org/pc/sif_interaction_rules.do">Simple Interaction Format (SIF)</a>. 
          A quick introduction to SBGN can be found <a target="_blank" href="https://sbgn.github.io/learning">here</a>. Please 
          refer to <a target="_blank" href="https://sbgn.github.io/referencecards">reference cards</a>, 
          <a target="_blank" href="https://sbgn.github.io/learnerscards">learner's cards</a> and 
          <a target="_blank" href="http://sbgn.github.io/sbgn/specifications">specifications</a> for details of SBGN languages. 
        </p>

        <p> Newt can be launched with a blank model, where a map can be created from scratch or an existing one can be uploaded 
          for editing. Some samples are also available under the File menu. Samples include SBGN PD and AF learner's cards as well. 
          In addition, one can launch Newt with a remote 
          <a href="https://web.newteditor.org/?URL=https://reactome.org/ContentService/exporter/event/R-HSA-450531.sbgn&inferNestingOnLoad=true&mapColorScheme=opposed_red_blue&fitLabelsToNodes=true" target="_blank">SBGN</a>, 
          <a href="https://web.newteditor.org/?URL=https://www.ebi.ac.uk/biomodels/model/download/BIOMD0000000125.2?filename=BIOMD0000000125_url.xml&inferNestingOnLoad=true&applyLayoutOnURL=true" target="_blank">SBML</a>, 
          <a href="https://web.newteditor.org/?URL=https://www.wikipathways.org/wikipathways-assets/pathways/WP121/WP121.gpml" target="_blank">GPML</a>, or
          model specified as a URL or as a 
          <a href="http://www.pathwaycommons.org/pc2/" target="_blank">Pathway Commons URI</a> 
          (<a href="http://web.newteditor.org/?URI=http://identifiers.org/reactome/R-HSA-6803211" target="_blank">example</a>).

        <p> <a href="tutorials-basics.html">Here</a> is a short tutorial illustrating basics of Newt. Also refer to "Help | 
          {Quick Help,Quick Tutorial,UI Guide}" for further help.</p>

        <p> A video illustrating version 4 features can be found <a href="tutorials-newt4.html">here<a>. </p>

        <h2 id="menubar-and-toolbar">Menubar and toolbar
          <a href="#top-of-tutorials"><img src="images/back-to-top.svg" height=16px/></a></h2>

        <p> Menubar located at the top of the Newt editor is where all capabilities of the tool are assembled in an organized 
          manner. Right below is a toolbar, which compiles more frequently used of all available operations for convenience. </p>

        <h2 id="right-panel">Right panel
          <a href="#top-of-tutorials"><img src="images/back-to-top.svg" height=16px/></a>
        </h2>

        <p> The panel on the right is composed of the following tabs. </p>

        <h3 id="inspector">Object properties inspector and editor
          <a href="#top-of-tutorials"><img src="images/back-to-top.svg" height=16px/></a>
        </h3>
        <span class="left"><a href="images/inspector.png" data-lightbox="image-inspector" data-title="Editing through the object inspector"><img src="images/inspector.png" style="width:400px"/></a></span>
        <p> The object inspector on the right panel titled "Object" is activated when a particular object is selected by left-clicking on it, enabling inspection as well as editing of the style of the underlying object from border color to font type and size to state variables and custom annotations.</p>
        <p> &#9752; When multiple objects are selected, the inspector will only display the common attributes (and their values if all objects have the same value for that attribute) and will allow convenient simultaneous modification of those attributes. </p>
	      <p> A convenient way to change the style of an entire map is to use one of pre-defined color schemes (see section on <a href="#color-scheme">Color Scheme</a>) under the Map tab in the right panel. 3D or gradient coloring of map objects are available through these colors schemes via background images as well as solid fill colors.</p>
        <p> Depending on the type of the object being inspected, a section for additional properties of the object from external web resouces may be presented as follows:
          <ul>
          <li> When the object is a <i>macromolecule</i> or a similar entity, a section named "Properties from GeneCards" will appear. If the user clicks on the button titled with the symbol of the entity available under this section, the corresponding page in <a href="https://www.genecards.org/" target="_blank">GeneCards</a>, if any, will be opened in a separate tab. </li>
          <li> When the object is a <i>simple chemical</i>, a section named "Properties from ChEBI" will appear. If the user unfolds this section, properties of the associated simple chemical from <a href="https://www.ebi.ac.uk/chebi/" target="_blank">CheBI</a>, if any, will be shown in-place in the inspector tab. </li>
          </ul>
        </p>
        <p> A section titled "Custom Properties" in the Object tab is dedicated to annotations of map objects. Here the user can add an annotation of a selected object using a <a href="https://co.mbine.org/standards/qualifiers">BioModels qualifier</a> or a custom key-value pair.
        </p>

        <h3 id="map-properties">Map properties
          <a href="#top-of-tutorials"><img src="images/back-to-top.svg" height=16px/></a>
        </h3>

        <span class="right"><a href="images/map-properties.png" data-lightbox="image-map-properties" data-title="Configuring the map through settings"><img src="images/map-properties.png" style="width:360px"/></a></span>
        <p> The tab titled "Map" on the right panel is for configuring the drawing of the map. It can be enabled with "File | Map Properties..." or through the associated tool on the toolbar. </p>
        <p> Numerous settings of a map from general properties like compound paddings to label related settings such as whether to fit labels to nodes, to color scheme can be set using this tab. </p>
        <p> Name and description of the map can also be set using this panel. </p>
        <p> &#9752; Notice that some of these settings apply to all map types, whereas others only apply to a specific map type. </p>
        <p> Type of a map can be one of the following:
          <ul>
            <li> <b>PD</b>: A <i>process description</i> map containing nodes/edges <i>only</i> from PD palette. A semantic validation check (Edit | Validate Map...) can be performed only when the map type is PD.
            <li> <b>AF</b>: An <i>activity flow</i> map containing nodes/edges <i>only</i> from AF palette
            <li> <b>PD+AF</b>: An <i>SBGN</i> map containing nodes/edges from PD or AF palette
            <span class="right"><a href="images/map-properties-tooltip.png" data-lightbox="image-map-properties" data-title="Quick help for settings via tooltips"><img src="images/map-properties-tooltip.png" style="width:240px"/></a></span>
            <li> <b>SIF</b>: A <i>SIF</i> map containing nodes/edges <i>only</i> from SIF palette
            <li> <b>SBML</b>: An <i>SBML</i> map containing nodes/edges from SBML palette
            <li> <b>PD+AF+SIF+SBML</b>: A <i>hybrid</i> map containing nodes/edges from any palette
          </ul>
        </p>
        <p> &#9752; When you are over a particular setting in this tab, a tooltip will show an explanation of the associated setting. </p>

        <h3 id="experiment-data">Experiment data
          <a href="#top-of-tutorials"><img src="images/back-to-top.svg" height=16px/></a>
        </h3>

        <p> One particular part of this tab that might be interest to most users is for uploading and showing experiment data in the context of your pathways.</p> 
        <span class="left"><a href="images/map-properties-experiment.png" data-lightbox="image-map-properties" data-title="Support for overlaying experiment data over the pathway"><img src="images/map-properties-experiment.png" style="width:440px"/></a></span>
        <p> Experiment data files are simple tab-delimited files with possibly multiple studies. Once loaded using "Experiment | Load..." menu or the + icon in the Experiment section of the Map tab, the background of each associated node whose label matches those given in the file will be colored according to the data value. Each file might contain one or more studies, and multiple files may be uploaded for analysis. The node background is split into as many times as studies available in all currently loaded and unhidden study. 
        </p>  
        <p>Users may inspect the uploaded experiment in a number of ways. First, it's color coded in the associated section of the node background. Secondly, when the mouse is over a particular node with some experiment data, a tooltip is displayed to show the corresponding values. Finally, a popup table is available through the table icon in the Experiment section of the Map tab. 
        </p>
        <p> &#9752; The values table is searchable by node label and could be sorted by node label or by values of a particular experiment by clicking on the associated column header. Another nice feature for locating the associated map object is by hovering a particular row, which will highlight the corresponding object in the map. </p>
        <p> Each file or study may be temporarilly hidden or unhidden by simply clicking on the file or study name, respectively. </p>
        <p> Below is a <a href="samples/sample-experiment-file.txt">sample <i>tab-delimited</i> experiment data file</a> for <a href="https://github.com/iVis-at-Bilkent/newt/blob/master/app/samples/tp53_rb_pathway.nwt">this</a> SIF map:</p>
          <div style="margin-left: 20px;"><code>
            version<span id="tab">\t</span>1.0<br>
            name<span id="tab">\t</span>sample experiment data<br>
            description<span id="tab">\t</span>Adenoid Cystic Carcinoma 2014 vs 2019<br>
            color<span id="tab">\t</span>0<span id="tab">\t</span>#FFFFFF<span id="tab">\t</span>100<span id="tab">\t</span>#FF0000<br>
            gene<span id="tab">\t</span>2014<span id="tab">\t</span>2019<br>
            RB1<span id="tab">\t</span>36<span id="tab">\t</span>12<br>
            TP53<span id="tab">\t</span>36<span id="tab">\t</span>72<br>
            CDKN2A<span id="tab">\t</span>0<span id="tab">\t</span>14<br>
            MDM2<span id="tab">\t</span>0<span id="tab">\t</span>5<br>
            CCNE<span id="tab">\t</span>0<span id="tab">\t</span>7<br><br>
          </code></div>
        <p>
          where:
          <ul>
          <li>The first optional line specifies the <b>version</b> of the file.</li>
          <li>The second optional line specifies the <b>name</b> of the experiment.</li>
          <li>The third optional line provides a <b>description</b> of the experiment.</li>
          <span class="right"><a href="images/map-properties-experiment-color.png" data-lightbox="image-map-properties" data-title="Example of color specification"><img src="images/map-properties-experiment-color.png" style="width:360px"/></a></span>
          <li>The fourth optional line provides a <b>color mapping</b> for values in the experiment. Here the goal is to map certain control points / values to colors. When these values are sorted in ascending order, anything outside the range is considered invalid and displayed as gray. Values that map to these exact control values are displayed in the color specified. And the values that fall in-between these control points are linearly mapped to the in-between color. For instance, for the color description provided below, values are restricted to be in range [-100, 100] with control points and corresponding colors being {-100, 0, 100} and {red, green, blue}, respectively, yielding a coloring as shown in the figure on the right:
            <div style="margin-left: 20px;"><code>
              color	-100	#FF0000	0	#00FF00	100	#0000FF<br><br>
            </code></div>
          &#9752; By default colors are expected to be in range [-100, 100] with colors blue and red, respectively at end points, and white at 0 (i.e. blue to white to red uniformly).
          </li>
          <li>The fifth mandatory line specifies the <b>names of the studies</b> in the experiment.</li>
          <li>The lines after the names of the studies will list the <b>data values</b> for each node with the given label in each study (one per column) in the experiment.</li>
          </ul>
        </p>

        <h3 id="color-scheme">Color scheme
          <a href="#top-of-tutorials"><img src="images/back-to-top.svg" height=16px/></a>
        </h3>

        <span class="right"><a href="images/color-scheme.png" data-lightbox="image-map-properties" data-title="Many pre-defined color schemes with solid, gradient, and 3D style are available"><img src="images/color-scheme.png" style="width:320px"/></a></span>
        <p> Newt provides a number of pre-defined color schemes with varying styles: <b>solid</b>, <b>gradient</b> or <b>3D</b> available under the Color Scheme section in the Map tab. The chosen color scheme is not only applied to all nodes/edges in the map but also to any new nodes and edges created interactively through the Palette. The icon to the right of each color scheme is for reversing the colors on that particular scheme. </p>
        <p> &#9752; Notice however that when experiment data overlay feature is in use, this section is disabled and the color scheme defaults to solid black and white as the background of nodes are used for showing color coded experiment values.</p>

        <h3 id="palette">Palette
          <a href="#top-of-tutorials"><img src="images/back-to-top.svg" height=16px/></a>
        </h3>
        <span class="left"><a href="images/palette.png" data-lightbox="image-pallette" data-title="Constructing a map using the palette"><img src="images/palette.png" style="width:440px"/></a></span>
        <p> When the Palette tab on the right panel is chosen or when the user clicks in an empty area on the drawing canvas, 
          the palette composed of nodes and edges of available map types is displayed. The type of a map (see Type under 
          General Properties) is considered PD by default when the user constructs a map from scratch, using only nodes and edges from the PD palette. The user, however, may change this
          on an empty map and use the corresponding palette. If the user wants to mix up nodes and/or edges from multiple palettes though, 
          they should set the map type to be a <i>PD+AF</i> (any SBGN node/edge) or <i>PD+AF+SIF+SBML</i> (any node/edge). 
          The notation specific checks (e.g. a modulation arc can only go from a source EPN such as macromolecule to a process node in 
          a PD map) are no longer performed for hybrid maps though. </p>

        <p> Nodes can either be created by dragging from the palette onto the canvas or by selecting a particular node on the palette and then clicking on the canvas. Clicking on a node or an edge puts you in node/edge creation mode, and upon creation of the node/edge, Newt returns to selection mode. Clicking <i>twice</i> on a tool, on the other hand, puts you in <i>sustained mode</i> so that the user can create multiple objects of that type in a row. </p>

	      <p> A convenient way to reconnect an edge is by selecting the edge and grabbing it from its reconnection handles at both ends and dragging onto its new source / target. </p>

        <h2 id="file-operations">File operations
          <a href="#top-of-tutorials"><img src="images/back-to-top.svg" height=16px/></a></h2>
        <p> Newt can read and write an extended version of <a href="https://github.com/sbgn/sbgn/wiki/SBGN_ML">SBGNML</a> files with ".nwt" extension. Extension includes style of map objects (such as color, font, and optional background image), custom properties (e.g. Uniprot ID associated with a macromolecule), and hidden/collapsed map content; thus, such information is restored upon save-reload. </p>

        <h3 id="import-export">Import-export
          <a href="#top-of-tutorials"><img src="images/back-to-top.svg" height=16px/></a>
        </h3>
        <p> If you'd like a version without such style and annotation information to perhaps exchange maps with other tools such as <a href="https://immersive-analytics.infotech.monash.edu/vanted/addons/sbgn-ed/">SBGN-ED</a>, however, you could get your map in plain SBGNML using "Export | SBGN-ML Plain". </p>

        <p> &#9752; Newt can support both milestones 0.2 and 0.3 but opts to 0.2 by default. Persistency in milestone 0.3 is available though through the "Export" menu.
        </p>

        <p> Newt also supports import and export facilities from and to <a href="https://en.wikipedia.org/wiki/SBML">SBML</a>, <a href="http://www.celldesigner.org/">CellDesigner</a>, and <a href="https://pathvisio.org/documentation/GPML">GPML</a> through "{Import, Export} | {SBML, CellDesigner, GPML}".</p>

        <p> In addition, AF models can be imported to Newt from simple tab-delimited files. An example model formatted in this manner follows:</p>
        <div style="margin-left: 20px;"><code>
        SBGNViz AF 1.0<br>
        Sample Pathway title<br>
        This is a sample pathway description.<br>
        --NODE_NAME<span id="tab">\t</span>NODE_ID<span id="tab">\t</span>NODE_TYPE<span id="tab">\t</span>PARENT_ID<span id="tab">\t</span>POSX<span id="tab">\t</span>POSY<span id="tab">\t</span>WIDTH<span id="tab">\t</span>HEIGHT--<br>
        ERBB2<span id="tab">\t</span>1<span id="tab">\t</span>BIOLOGICAL_ACTIVITY<span id="tab">\t</span>-1<br>
        NRAS<span id="tab">\t</span>2<span id="tab">\t</span>BIOLOGICAL_ACTIVITY<span id="tab">\t</span>-1<br>
        ...<br>
        --EDGE_ID<span id="tab">\t</span>SOURCE<span id="tab">\t</span>TARGET<span id="tab">\t</span>EDGE_TYPE--<br>
        11<span id="tab">\t</span>1<span id="tab">\t</span>2<span id="tab">\t</span>NEGATIVE_INFLUENCE<br>
        ...
        </code></div>

        <p><br>Furthermore, maps may be imported from or exported to <a target="_blank" href="https://www.pathwaycommons.org/pc/sif_interaction_rules.do">Simple Interaction Format (SIF)</a> files using the "{Import,Export} | SIF" File menu items. A SIF file is formatted as follows:</p>
        <div style="margin-left: 20px;"><code>
        entity_id<span id="tab">\t</span>relationship_type<span id="tab">\t</span>entity_id
        ...
        </div></code>

        <p><br>where <code>entity_id</code> is the label and ID of a macromolecule or a simple chemical and <code>relationship_type</code> is one of the predetermined types of relationships between these two entities.
        </p>

      	<p> Here is a sample SIF file along with sample style and layout files, respectively (<a href="samples/Platelet-GP6-activation.sif">SIF</a>, <a href="samples/Platelet-GP6-activation.format">style</a> and <a href="samples/Platelet-GP6-activation.layout">layout</a>).    
	      </p>

        <p> Below is a table showing how each import or export operation behaves. Note that the current map type matters in most such operations.
        <span class="left"><a href="images/import-export-table.png" data-lightbox="image-pallette" data-title="Table showing behavior of each import/export operation"><img src="images/import-export-table.png" style="width:100%"/></a></span>
        </p>
        
        <h3 id="static-images">Static images
          <a href="#top-of-tutorials"><img src="images/back-to-top.svg" height=16px/></a>
        </h3>

        <p> File menu can be used to export the current map as a static image (either as a raster file in PNG or JPG or as scalable vector graphics in SVG).</p>

        <h3 id="persisting-preferences">Persisting user preferences
          <a href="#top-of-tutorials"><img src="images/back-to-top.svg" height=16px/></a>
        </h3>

	<p>Users might like to save their preferences (including Map, Layout, Grid, and Object properties) for later use using "File | Save Preferences..." in a file with ".newtp" extension. These preferences can later be loaded using "File | Load Preferences..." to be applied to any map.</p>	

        <h3 id="remote-launch">Launch with a remote SBGN map
          <a href="#top-of-tutorials"><img src="images/back-to-top.svg" height=16px/></a>
        </h3>

	<p>As noted above Newt may be launched by a remote file via a URL (<a
    href="http://web.newteditor.org/?URL=https://raw.githubusercontent.com/sbgn/libsbgn/master/test-files/PD/activated_stat1alpha_induction_of_the_irf1_gene.sbgn&mapColorScheme=blue_scale"
    target="_blank">example</a>) or by URI (<a
    href="http://web.newteditor.org/?URI=http://identifiers.org/reactome/R-HSA-6803211&mapColorScheme=opposed_green_brown"
    target="_blank">example</a>) as a query to the Pathway Commons database. In the event that the provided URL is a file containing
    a non-native Newt format (e.g., GPML), Newt will automatically apply a conversion and display the provided model as an SBGN map. The user 
    might like to change some map properities on a remote launch. In case a map property is also specified 
    inside the file, they will be overwritten by these parameters passed during remote launch. In the case of 
    <code>compoundPadding</code> and <code>inferNestingOnLoad</code>, however, any values specified inside the file will be ignored. 
    Available properties (with default values listed first) include:
    <ul>
    <li><code>compoundPadding=0</code>,
    <li><code>inferNestingOnLoad=false</code>,
    <li><code>arrowScale=1.25</code>,
    <li><code>showComplexName=true</code>,
    <li><code>dynamicLabelSize=regular</code>,
    <li><code>fitLabelsToNodes=false</code>,
    <li><code>fitLabelsToInfoboxes=false</code>,
    <li><code>recalculateLayoutOnComplexityManagement=true</code>,
    <li><code>rearrangeOnComplexityManagement=true</code>,
    <li><code>animateOnDrawingChanges=true</code>,
    <li><code>adjustNodeLabelFontSizeAutomatically=false</code>,
    <li><code>enablePorts=true</code>,
    <li><code>allowCompoundNodeResize=false</code>,
    <li><code>mapColorScheme={black_white, black_white, greyscale, inverse_greyscale, blue_scale, inverse_blue_scale, opposed_red_blue, opposed_red_blue2, opposed_green_brown, opposed_green_brown2, opposed_purple_brown, opposed_purple_brown2, opposed_purple_green, opposed_purple_green2, opposed_grey_red, opposed_grey_red2, pure_white}</code>,
    <li><code>mapName=""</code>,
    <li><code>mapDescription=""</code>,
    <li><code>enableSIFTopologyGrouping=false</code>,
    <li><code>applyLayoutOnURL=false</code>. 
</p>

        <h2 id="convenient-editing">Convenient editing via SBGN bricks
          <a href="#top-of-tutorials"><img src="images/back-to-top.svg" height=16px/></a></h2>

        <p> Map content can be expanded using the Palette as described earlier. A convenient way to populate a map is using the templates under the Edit (Add Using SBGN Bricks..) menu inspired by <a href="https://sbgnbricks.org/">SBGN bricks</a>. </p>

        <table class="plaintable">
          <tr>
            <th class="leftcolumn">
            <a href="images/sbgn-bricks.png" data-lightbox="image-templates" data-title="Convenient editing using SBGN bricks"><img class="ssthumbnail" src="images/sbgn-bricks.png" style="height: 240px;"></a>
            </th>
          </tr>
        </table>

        <h2 id="alignment-grid-and-guidelines">Alignment, grid and guidelines
          <a href="#top-of-tutorials"><img src="images/back-to-top.svg" height=16px/></a></h2>

        <p> Newt provides various mechanisms for alignment of map objects. One such facility is available through the View menu, where the user can select a number of objects and align their center or edge horizontally or vertically. Such alignment is performed with respect to the object selected first. </p>

        <p> Another way to perform alignment is through the grid. When you enable the grid tool (<img src="images/enable-grid.svg" title="Tool for enabling the grid" style="height: 16px;">) on the toolbar, a grid is displayed and nodes are snapped to grid as they are dragged. </p>

        <span class="right"><a href="images/guideline-example.png" data-lightbox="image-grid-guide" data-title="An example where the dragged node is snapped to an alignment position"><img src="images/guideline-example.png" style="width: 460px;"/></a></span>
        <p> Yet another way to perform alignment is by enabling the guidelines with the associated tool on the toolbar (<img src="images/enable-guidelines.svg" title="Tool for enabling the guidelines" style="height: 16px;">). Two types of guidelines are enabled with this tool:
<ul>
 <li> Geometric guidelines (e.g. center of two nodes are in alignment) and
 <li> Distance guidelines (e.g. a process node is distanced equally from its source and target).
</ul>
Grid and guideline support is rather involved and can be customized using the settings through the View menu.
</p>

        <h2 id="semantic-validation"> Semantic validation
          <a href="#top-of-tutorials"><img src="images/back-to-top.svg" height=16px/></a></h2>

	<span class="right"><a href="images/validation-example.png" data-lightbox="image-validation-example" data-title="An example where the user is suggested a fix for a particular error"><img src="images/validation-example.png" style="width: 400px;"/></a></span>
	<p> SBGN PD maps may be validated using "Edit | Validate Map..." menu item. This will validate the map and show the user any semantic problems with the map in a newly created tab titled "Console" in the right panel. Whenever possible, Newt will suggest a fix for each particular problem, allowing the user to choose from potentially multiple ways to fix the problem in this new tab.  </p>

        <h2 id="complexity-management">Complexity management
          <a href="#top-of-tutorials"><img src="images/back-to-top.svg" height=16px/></a></h2>

        <p> Some maps can be rather large. No matter how advanced of layout and viewing facilities are available, navigation through such maps and tracing interactions can be cumbersome. Newt provides some advanced complexity management techniques [<a href="#references">2</a>,<a href="#references">1</a>,<a href="#references">3</a>] so that the user can focus on a sub-map of interest at any time during their analysis, removing unwanted details <i>temporarily</i> from the map. Below we discuss these facilities. </p>

        <p> <a href="tutorials-compl.html">Here</a> is a short tutorial illustrating complexity management in Newt. </p>

        <h3 id="hide-show">Hide-show
          <a href="#top-of-tutorials"><img src="images/back-to-top.svg" height=16px/></a>
        </h3>
        <p> Unwanted details of a map may be hidden by the user temporarily through the View menu. Below is an example, where the user initially keeps a single process and hides all others (top). Then the user gradually reveals the hidden neighbors (bottom maps) using "Show Hidden Neighbors" in the node context menu. </p>

        <table class="plaintable">
          <tr>
            <th colspan="4">
            <a href="images/hide-example-1.png" data-lightbox="image-hide" data-title="Map content is hidden except for a process"><img class="ssthumbnail" src="images/hide-example-1.png" style="height: 360px;"/></a>
            </th>
          </tr>
          <tr>
            <th>
            <a href="images/hide-example-2.png" data-lightbox="image-hide" data-title="Then the user starts to gradually show hidden content"><img class="ssthumbnail" src="images/hide-example-2.png" style="height: 70px;"/></a>
            </th>
            <th>
            <a href="images/hide-example-3.png" data-lightbox="image-hide" data-title="One more level of neighborhood is revealed"><img class="ssthumbnail" src="images/hide-example-3.png" style="height: 86px;"/></a>
            </th>
            <th>
            <a href="images/hide-example-4.png" data-lightbox="image-hide" data-title="Unhide operation continues as deep as desired"><img class="ssthumbnail" src="images/hide-example-4.png" style="height: 100px;"/></a>
            </th>
            <td>...</td>
          </tr>
        </table>

        <h3 id="expand-collapse">Expand-collapse
          <a href="#top-of-tutorials"><img src="images/back-to-top.svg" height=16px/></a>
        </h3>
        <p> Another way to manage complexity is by <i>collapsing</i> currently not-so-interesting parts of a map (i.e. a submap, a compartment or a complex). Collapsed compound node can later be <i>expanded</i> to reveal its nested content. </p>

        <p> Below is an example, where, first, all complexes of a map have been collapsed to reduce complexity. Then, nucleus was collapsed to focus on the rest of the map. </p>
        <table class="plaintable">
          <tr>
            <th colspan="2">
            <a href="images/expand-collapse-example-1.png" data-lightbox="image-expand" data-title="Original graph is composed of many complexes with members shown, and a couple of cellular compartments"><img class="ssthumbnail" src="images/expand-collapse-example-1.png" style="height: 240px;"/></a>
            </th>
          </tr>
          <tr>
            <th>
            <a href="images/expand-collapse-example-2.png" data-lightbox="image-expand" data-title="After all complexes are collapsed"><img class="ssthumbnail" src="images/expand-collapse-example-2.png" style="height: 220px;"/></a>
            </th>
            <th>
            <a href="images/expand-collapse-example-3.png" data-lightbox="image-expand" data-title="After nucleus is collapsed"><img class="ssthumbnail" src="images/expand-collapse-example-3.png" style="height: 132px;"/></a>
            </th>
          </tr>
        </table>

        <h2 id="search-and-highlight">Search and highlight
          <a href="#top-of-tutorials"><img src="images/back-to-top.svg" height=16px/></a></h2>
        <p> Newt allows users to highlight a specified part of a map through the Highlight menu (e.g. Highlight | Selected). To highlight the part of a map involving a certain object, you may use the Search facility (Highlight | Search). Newt looks for node labels with specified keyword and highlights them (all processes involved in case of map type being PD). </p>
        <span class="center"><a href="images/search-example.png" data-lightbox="image-search" data-title="Part of the map containing &quot;fructose&quot; in its label is searched for and highlighted"><img src="images/search-example.png" style="width: 500px;" /></a></span>

        <h2 id="automatic-layout">Automatic layout
          <a href="#top-of-tutorials"><img src="images/back-to-top.svg" height=16px/></a></h2>

        <span class="left"><a href="images/before-layout.svg" data-lightbox="image-layout" data-title="A map with random layout"><img src="images/before-layout.svg" style="height: 220px;" /></a></span>
        <span class="right"><a href="images/after-layout.svg" data-lightbox="image-layout" data-title="After Newt's automatic layout was performed on the same map"><img src="images/after-layout.svg" style="height: 220px;" /></a></span>

        <p> Newt provides automatic layout services to lay out a map from scratch (disable Incremental flag under Layout | Layout Properties...) or incrementally (default). Newt uses a layout algorithm [<a href="#references">5</a>,<a href="#references">4</a>,<a href="#references">6</a>] respecting compound structures (e.g. compartments). Please refer to the <a target="_blank" href="https://github.com/iVis-at-Bilkent/cytoscape.js-fcose">related extension</a> for options for configuring automatic layout. </p>

        <p> The same algorithm is also applied incrementally after complexity management operations described earlier, if the associated flag (Map Properties... | Arrangement | Rearrange on Complexity Management) is enabled. </p>

        <p> <a href="tutorials-layout.html">Here</a> is a short tutorial illustrating automatic layout capabilities of Newt. </p>

        <h2 id="querying-pathway-databases">Querying Pathway Databases <a href="#top-of-tutorials"><img src="images/back-to-top.svg" height=16px/></a></h2>

        <p> Newt provides an extensive interface to perform live queries to the <a target="_blank" href="http://pathwaycommons.org/">Pathway 
          Commons</a> database, presenting results in SBGN, and enabling editing on them. Currently available type of queries 
          include:</p>
		<ul>
		 <li> Query Paths Between (find all paths of specified length limit between any two entities of the specified set),
		 <li> Query From To.. (find all paths of specified length limit from a set of source entities to a set of target entities),
		 <li> Common Stream (find molecules within the specified distance that are affecting or affected by a specific set of molecules in common),
		 <li> Neighborhood (find the neighbors of an entity within a certain distance), and
		 <li> Query by URI
		</ul> 
	<p>as detailed <a target="_blank" href="http://www.pathwaycommons.org/pc2/">here</a> [<a href="#references">7</a>].</p>

  <span class="left"><a href="images/PC-query-dialog.png" data-lightbox="image-pc-query" data-title="Sample Paths Between query"><img src="images/PC-query-dialog.png" style="height: 190px;" /></a></span>

	<p>For instance, Paths Between query takes a list of gene symbols (separated by blank spaces or on separate lines) and a length 
    limit specifying the maximum length of paths between specified genes. Since there could be exponentially many paths between 
    multiple genes, this length limit can be at most 3.</p> 

  <p>Additional options are used to enable removal of disconnected entities and
    redundant processes as there might be duplicate instances of the same process due to Pathway Commons gathering pathways from
    multiple different databases. </p>

    <span class="right"><a href="images/PC-query-result.png" data-lightbox="image-pc-query" data-title="Partial view of the result of sample query"><img src="images/PC-query-result.png" style="height: 390px;" /></a></span>

  <p>Query results are highlighted, where source and/or target genes are colored in green in this example and the connecting paths
    are in orange.
  </p>

  <p>Newt also provides querying <a href="https://reactome.org/">Reactome</a> by Reactome ID, and 
    querying <a href="https://www.wikipathways.org/">WikiPathways</a> by WPID.</p>

        <h2 id="references">References
          <a href="#top-of-tutorials"><img src="images/back-to-top.svg" height=16px/></a></h2>

        [1] H. Balci, M.C. Siper, N. Saleh, I. Safarli, L. Roy, M. Kilicarslan, R. Ozaydin, A. Mazein, C. Auffray, O. Babur, E. Demir and U. Dogrusoz, <a href="https://doi.org/10.1093/bioinformatics/btaa850">Newt: a comprehensive web-based tool for viewing,constructing, and analyzing biological maps</a>, Bioinformatics, 37(10), pp. 1475–1477, 2021.
        <br><br>[2] M. Sari, I. Bahceci, U. Dogrusoz, S.O. Sumer, B.A. Aksoy, O. Babur, E. Demir, "<a href="https://doi.org/10.1371/journal.pone.0128985">SBGNViz: a tool for visualization and complexity management of SBGN process description maps</a>", PLoS ONE, 2015. 
        <br><br>[3] U. Dogrusoz and B. Genc, "<a href="https://doi.org/10.1016/j.cag.2005.10.015">A Multi-Graph Approach to Complexity Management in Interactive Graph Visualization</a>", Computers & Graphics, 2006.
        <br><br>[4] U. Dogrusoz, A. Karacelik, I. Safarli, H. Balci, L. Dervishi, M.C. Siper, "<a href="https://doi.org/10.1371/journal.pone.0197238">Efficient methods and readily customizable libraries for managing complexity of large networks</a>", PLoS ONE, 2018.
        <br><br>[5] B. Genc, U. Dogrusoz, "<a href="https://doi.org/10.1093/bioinformatics/btv516">An Algorithm for Automated Layout of Process Description Maps Drawn in SBGN</a>", Bioinformatics, 32(1), pp. 77-84, 2016.
        <br><br>[6] U. Dogrusoz, E. Giral, A. Cetintas, A. Civril, and E. Demir, "<a href="https://doi.org/10.1016/j.ins.2008.11.017">A Layout Algorithm For Undirected Compound Graphs</a>", Information Sciences, 2009.
        <br><br>[7] U. Dogrusoz, E. Giral, A. Cetintas, A. Civril, and E. Demir, "A Compound Graph Layout Algorithm for Biological Pathways", LNCS, vol. 3383, pp. 442-447, 2004.
        <br><br>[8] U. Dogrusoz, A. Cetintas, E. Demir, and O. Babur, "<a href="https://doi.org/10.1186/1471-2105-10-376">Algorithms for Effective Querying of Compound Graph-Based Pathway Databases</a>", BMC Bioinformatics, 10:376, 2009. 
 
      </div>
    </div>
    <script> writeFooter(); </script>
  </div>
</body>
</html>
